mutiny-wasm/src/indexed_db.rs

```diff
@@ -0,0 +1,498 @@
+use anyhow::anyhow;
+use gloo_utils::format::JsValueSerdeExt;
+use mutiny_core::error::{MutinyError, MutinyStorageError};
+use mutiny_core::storage::MutinyStorage;
+use rexie::{ObjectStore, Rexie, TransactionMode};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+use wasm_bindgen::JsValue;
+use wasm_bindgen_futures::spawn_local;
+
+pub(crate) const WALLET_DATABASE_NAME: &str = "wallet";
+pub(crate) const WALLET_OBJECT_STORE_NAME: &str = "wallet_store";
+
+#[derive(Clone)]
+pub struct IndexedDbStorage {
+    pub(crate) password: Option<String>,
+    /// In-memory cache of the wallet data
+    /// This is used to avoid having to read from IndexedDB on every get.
+    /// This is a RwLock because we want to be able to read from it without blocking
+    memory: Arc<RwLock<HashMap<String, Value>>>,
+    pub(crate) indexed_db: Arc<RwLock<Option<Rexie>>>,
+}
+
+impl IndexedDbStorage {
+    pub async fn new(password: Option<String>) -> Result<IndexedDbStorage, MutinyError> {
+        let indexed_db = Arc::new(RwLock::new(Some(Self::build_indexed_db_database().await?)));
+
+        let map = Self::read_all(&indexed_db).await?;
+        let memory = Arc::new(RwLock::new(map));
+
+        Ok(IndexedDbStorage {
+            password,
+            memory,
+            indexed_db,
+        })
+    }
+
+    async fn save_to_indexed_db(
+        indexed_db: &Arc<RwLock<Option<Rexie>>>,
+        key: &str,
+        data: &Value,
+    ) -> Result<(), MutinyError> {
+        let tx = indexed_db
+            .try_write()
```

Comment:
Im curious about how the `indexed_db.try_write()` works with `tx.done()` and when to actually get the write lock. If we're just saving to `indexed_db`, do we need the write lock here? Is something being mutated?

I'm wondering if we only need the `write` lock if we are actually changing `Rexie`, IE, whenever it is being stopped or swapped out with another `Rexie`

Reply:
Yeah I am not too sure, I think you added the rw lock around rexie with the start stop stuff.

Reply:
Yeah that's true. Doesnt hurt things and is just a copy but fine to leave it for now.